File: platform_src/android/app/src/main/java/org/elastos/trinity/runtime/PreferenceManager.java
package org.elastos.trinity.runtime;

import android.content.Context;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.content.res.AssetManager;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.InputStream;
import java.util.Iterator;
import java.util.Locale;

public class PreferenceManager {
    private static PreferenceManager preferenceManager;

    private JSONObject defaultPreferences = new JSONObject();
    ManagerDBAdapter dbAdapter = null;

    PreferenceManager() {
        dbAdapter = AppManager.getShareInstance().getDBAdapter();
        try {
            parsePreferences();
        } catch (Exception e) {
            e.printStackTrace();
        }
        PreferenceManager.preferenceManager = this;

        // Make some services ready
        UIStyling.prepare(true); // TMP - USE MASTER BRANCH CODE NOT THIS
    }

    public static PreferenceManager getShareInstance() {
        if (PreferenceManager.preferenceManager == null) {
            PreferenceManager.preferenceManager = new PreferenceManager();
        }
        return PreferenceManager.preferenceManager;
    }

    public void parsePreferences() throws Exception {
        AssetManager manager = AppManager.getShareInstance().activity.getAssets();
        InputStream inputStream = manager.open("www/config/preferences.json");

        defaultPreferences = Utility.getJsonFromFile(inputStream);
        defaultPreferences.put("version", getVersion());
    }

    private Object getDefaultValue(String key) throws Exception {
        Object value = null;
        if (defaultPreferences.has(key)) {
            value = defaultPreferences.get(key);
        }
        return value;
    }

    public JSONObject getPreference(String key) throws Exception  {
        Object defaultValue = getDefaultValue(key);
        if (defaultValue == null) {
            throw new Exception("getPreference error: no such preference!");
        }

        JSONObject ret = dbAdapter.getPreference(key);

        if (ret == null) {
            ret = new JSONObject();
            ret.put("key", key);
            ret.put("value", defaultValue);
        }

        if (key.equals("locale.language") && ret.getString("value").equals("native system")) {
            ret.put("value", Locale.getDefault().getLanguage());
        }

        return ret;
    }

    public JSONObject getPreferences() throws Exception {
        JSONObject values = dbAdapter.getPreferences();
        Iterator keys = defaultPreferences.keys();
        while (keys.hasNext()) {
            String key = (String)keys.next();
            if (!values.has(key)) {
                Object value = defaultPreferences.get(key);
                values.put(key, value);
            }

            if (key.equals("locale.language") && values.getString(key).equals("native system")) {
                values.put(key, Locale.getDefault().getLanguage());
            }
        }
        return values;
    }

    final static String[] refuseSetPreferences = {
            "version"
    };

    private Boolean isAllowSetPreference(String key) {
        for (String item : PreferenceManager.refuseSetPreferences) {
            if (item.equals(key)) {
                return false;
            }
        }
        return true;
    }

    public void setPreference(String key, Object value) throws Exception {
        if (!isAllowSetPreference(key)) {
            throw new Exception("setPreference error: " + key + " can't be set!");
        }

        Object defaultValue = getDefaultValue(key);
        if (defaultValue == null) {
            throw new Exception("setPreference error: no such preference!");
        }

        if (dbAdapter.setPreference(key, value) < 1) {
            throw new Exception("setPreference error: write db error!");
        }

//        if (key == "developer.mode") {
//            Boolean isMode = false;
//            if (value != null) {
//                isMode = Boolean.getBoolean(value);
//            }
//
//            if (isMode) {
//                CLIService.getShareInstance().start();
//            }
//            else {
//                CLIService.getShareInstance().stop();
//            }
//        }

        JSONObject data = new JSONObject();
        data.put("key", key);
        data.put("value", value);
        JSONObject json = new JSONObject();
        json.put("action", "preferenceChanged");
        json.put("data", data);
        AppManager.getShareInstance().broadcastMessage(AppManager.MSG_TYPE_IN_REFRESH,
                json.toString() , "system");
    }

    public Boolean getDeveloperMode() {
        JSONObject value = null;
        Boolean ret = false;
        try {
            value = getPreference("developer.mode");
            if (value != null ) {
                ret = Boolean.getBoolean(value.getString("value"));
            }
        }
        catch (Exception e){
            e.printStackTrace();
        }


        return ret;
    }

    public void setDeveloperMode(Boolean value) {
        try {
            setPreference("developer.mode", value);
        }
        catch (Exception e){
            e.printStackTrace();
        }
    }

    public String getCurrentLocale() throws Exception {
        JSONObject value = getPreference("locale.language");
        String ret = value.getString("value");
        if (ret.equals("native system")) {
            ret = Locale.getDefault().getLanguage();
        }
        return ret;
    }

    public void setCurrentLocale(String code) throws Exception {
        setPreference("locale.language", code);
        JSONObject json = new JSONObject();
        json.put("action", "currentLocaleChanged");
        json.put("data", code);
        AppManager.getShareInstance().broadcastMessage(AppManager.MSG_TYPE_IN_REFRESH,
                json.toString(), AppManager.LAUNCHER);
    }

    public String getVersion() throws Exception {
        Context context = AppManager.getShareInstance().activity;
        PackageManager manager = context.getPackageManager();
        PackageInfo info = manager.getPackageInfo(context.getPackageName(), 0);
        return info.versionName;
    }

    public String getWalletNetworkType()  {
        String value = null;
        try {
            JSONObject item = getPreference("chain.network.type");
            value = item.getString("value");
        }
        catch (Exception e){
            e.printStackTrace();
        }

        if (value == null) {
            value = "MainNet";
        }

        return value;
    }

    public String getWalletNetworkConfig()  {
        String value = null;
        try {
            JSONObject item = getPreference("chain.network.config");
            value = item.getString("value");
        }
        catch (Exception e){
            e.printStackTrace();
        }

        if (value == null) {
            value = "";
        }

        return value;
    }

    public String getDIDResolver()  {
        String value = null;
        try {
            JSONObject item = getPreference("did.resolver");
            value = item.getString("value");
        }
        catch (Exception e){
            e.printStackTrace();
        }

        if (value == null) {
            value = "http://api.elastos.io:20606";
        }

        return value;
    }
}

File: platform_src/ios/Elastos/Classes/PreferenceManager.swift
 /*
  * Copyright (c) 2020 Elastos Foundation
  *
  * Permission is hereby granted, free of charge, to any person obtaining a copy
  * of this software and associated documentation files (the "Software"), to deal
  * in the Software without restriction, including without limitation the rights
  * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  * copies of the Software, and to permit persons to whom the Software is
  * furnished to do so, subject to the following conditions:
  *
  * The above copyright notice and this permission notice shall be included in all
  * copies or substantial portions of the Software.
  *
  * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  * SOFTWARE.
  */

import Foundation

//For system preference
 @objc(PreferenceManager)
 class PreferenceManager: NSObject {
    private static var preferenceManager: PreferenceManager?;
    private var defaultPreferences = [String: Any]();
    let dbAdapter: ManagerDBAdapter;

    override init() {
        dbAdapter = AppManager.getShareInstance().getDBAdapter();
        super.init();

        parsePreferences();
        PreferenceManager.preferenceManager = self;
    }

    @objc static func getShareInstance() -> PreferenceManager {
        if (PreferenceManager.preferenceManager == nil) {
            PreferenceManager.preferenceManager = PreferenceManager();
        }
        return PreferenceManager.preferenceManager!;
    }

    func parsePreferences() {
        do {
            let path = getAbsolutePath("www/config/preferences.json");
            defaultPreferences = try getJsonFromFile(path);
            defaultPreferences["version"] = getVersion();
        }
        catch let error {
            print("Parse preferences.json error: \(error)");
        }
    }

    private func anyToString(_ value: Any) -> String? {
        if (value is String) {
            return (value as! String);
        }
        else if (value is Bool) {
            return (value as! Bool).toString();
        }
        else if (value is [String]) {
            return (value as! [String]).description
        }
        else if (value is [String: Any]) {
            return (value as! [String: Any]).toString()!;
        }
        else if (value is Int) {
            return String(value as! Int)
        }
        else if (value is Double) {
            return String(value as! Double)
        }
        else if (value is NSNull) {
            return "null";
        }

        return nil;
    }

    private func getDefaultValue(_ key: String) -> Any? {
        let value = defaultPreferences[key];
        return value;
    }

    func getPreference(_ key: String) throws -> [String: Any] {
        let defaultValue = getDefaultValue(key);
        guard defaultValue != nil else {
            throw AppError.error("getPreference error: no such preference!");
        }

        var ret = try dbAdapter.getPreference(key);
        if (ret == nil) {
            ret = ["key": key, "value": defaultValue!] as [String : Any];
        }

        if (key == "locale.language" && ret!["value"] as? String == "native system") {
            ret![key] = getCurrentLanguage();
        }

        return ret!;
    }

    func getPreferences() throws -> [String: Any] {
        var values = try dbAdapter.getPreferences();
        for (key, value) in defaultPreferences {
            if (values[key] == nil) {
                values[key] = value;
            }
            if (key == "locale.language" && (values[key] as! String) == "native system") {
                values[key] = getCurrentLanguage();
            }
        }

        return values;
    }

    static let refuseSetPreferences = [
        "version"
    ];

    private func isAllowSetPreference(_ key: String) -> Bool {
        if (!PreferenceManager.refuseSetPreferences.contains(key)) {
            return true;
        }
        else {
            return false;
        }
    }

    @objc func setPreference(_ key: String, _ value: Any?) throws {
        if (!isAllowSetPreference(key)) {
            throw AppError.error("setPreference error: \(key) can't be set!");
        }

        let defaultValue = getDefaultValue(key);
        guard defaultValue != nil else {
            throw AppError.error("setPreference error: no such preference!");
        }

        try dbAdapter.setPreference(key, value);
        if (key == "developer.mode") {
            var isMode = false;
            if (value != nil && (value is Bool)) {
                isMode = value! as! Bool;
            }

            if (isMode) {
                CLIService.getShareInstance().start();
            }
            else {
                CLIService.getShareInstance().stop();
            }
        }

        let dict = ["action": "preferenceChanged", "data": ["key": key, "value": value]] as [String : Any];
        AppManager.getShareInstance().broadcastMessage(AppManager.MSG_TYPE_IN_REFRESH, dict.toString()!, "system");
    }

    func getDeveloperMode() -> Bool {
        let value = try? getPreference("developer.mode");
        var ret = false;
        if (value != nil && (value!["value"] is Bool)) {
            ret = value!["value"] as! Bool;
        }

        return ret;
    }

    func setDeveloperMode(_ value: Bool) {
        try? setPreference("developer.mode", value);
    }

    func getCurrentLocale() throws -> String {
        let value = try getPreference("locale.language");
        var ret = value["value"] as! String;
        if (ret == "native system") {
            ret = getCurrentLanguage();
        }
        return ret;
    }

    func setCurrentLocale(_ code: String) throws {
        let dict = ["action": "currentLocaleChanged", "code": code] as [String : Any];
        try setPreference("locale.language", code);
        AppManager.getShareInstance().broadcastMessage(AppManager.MSG_TYPE_IN_REFRESH, dict.toString()!, "launcher");
    }

    func getVersion() -> String? {
        let infoDictionary = Bundle.main.infoDictionary

        let majorVersion = infoDictionary? ["CFBundleShortVersionString"] as? String;

        return majorVersion;
    }

    @objc func getWalletNetworkType() -> String {
        var value: String? = nil;
        do {
            let item = try getPreference("chain.network.type");
            value = item["value"] as? String;
        }
        catch let error {
            print("getWalletNetworkType error: \(error)");
        }

        if (value == nil) {
            value = "MainNet";
        }
        return value!;
     }

    @objc func getWalletNetworkConfig() -> String {
        var value: String? = nil;
        do {
            let item = try getPreference("chain.network.config");
            if !(item["value"] is NSNull) {
                value = anyToString(item["value"]!);
            }
        }
        catch let error {
            print("getWalletNetworkConfig error: \(error)");
        }

        if (value == nil) {
            value = "";
        }
        return value!;
     }
}

